R/sparql.R

```diff
@@ -0,0 +1,176 @@
+#'       "UniRef90_A0A010Q9A4", "UniRef90_A0A010Q9B7", "UniRef90_A0A010Q9E1",
+#'       "UniRef90_A0A010Q9E5", "UniRef90_A0A010Q9E7", "UniRef90_A0A010Q9H0",
+#'       "UniRef90_A0A010Q9I7")
+#'
+#'
+#' SPARQLmap(x, .by = uniref ~ ec,  endpoint = "uniprot")
+#'
+#' SPARQLmap(x, .by = uniref ~ species,  endpoint = "uniprot")
+#'
+SPARQLmap <- function(
+        x, x.type = "uniref", .by = NULL, endpoint = "uniprot"
+        ) {
+
+    terms <- if (inherits(.by, "formula"))
+        all.vars(.by)
+    else .by
+
+    known_endpoints <- c("uniprot")
+    endpoint <- known_endpoints[charmatch(endpoint, table = known_endpoints)]
+    endpoint_url <- switch(endpoint, "uniprot" = "https://sparql.uniprot.org/")
+
+    query <- .composeSPARQL(x, x.type, terms, endpoint_url)
+    output <- .querySPARQL(query)
+    colnames(output) <- terms
+    return(output)
+}
+
+#' @importFrom utils URLencode
+.composeSPARQL <- function(x, x.type, terms, endpoint_url) {
+    #uniref.ids <- paste0("uniref:", x, collapse = " ")
+ terms_remote <- vapply(terms, .uniprot_terms, "")
+
+ prefixes <-
+"PREFIX up_core: <http://purl.uniprot.org/core/>
+PREFIX uniref: <http://purl.uniprot.org/uniref/>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+"
+
+ uniref2taxonomy <-
+"?unirefId up_core:member ?member .
+?member up_core:organism ?taxId .
+?taxId up_core:scientificName ?sciName .
+"
+
+ uniref2ec <-
+"?protein up_core:representativeFor ?unirefId .
+?protein ( up_core:enzyme | up_core:domain/up_core:enzyme | up_core:component/up_core:enzyme ) ?ecId .
+"
+
+q <- paste0("
+SELECT", paste0(terms_remote, collapse = " "), "
+WHERE {
+", .composeX(x, x.type), "
+",
+paste0(
+    if("?ecId" %in% terms_remote) uniref2ec,
+    if("?sciName" %in% terms_remote) uniref2taxonomy
+    ),
+"}")
+    # Build query
+    query <- paste0(prefixes, q)
+    query <- gsub("\\+", "%2B", utils::URLencode(query, reserved = TRUE))
+    query <- paste0(endpoint_url, "?query=", query)
+    return(query)
+}
+
+.composeX <- function(x, x.type) {
+    if(length(x) == 0L) return("")
+    if(x.type %in% c("uniref", "uniref50", "uniref90", "uniref100")) {
+        return(
+            paste0(
+                "VALUES ?unirefId {
+            ", paste0("uniref:", x, collapse = " "), "
+        } .")
+        )
+    }
+    if(x.type %in% c("ec")) {
+        return(
+            paste0(
+                "VALUES ?ecId {
+            ", paste0("enzyme:", x, collapse = " "), "
+        } .")
+        )
+    }
+    if(x.type %in% c("species")) {
+        return(
+            paste0(
+                "VALUES ?SciName {
+            ", paste0("up_core:", x, collapse = " "), "
+        } .")
+        )
+    }
+}
+
+#' @importFrom httr content GET timeout add_headers
+#' @importFrom utils URLencode read.csv
+#' @noRd
+.querySPARQL <- function(query) {
+
+    result <- httr::content(
+        httr::GET(
+            query,
+            httr::timeout(60),
```

Comment:
If this is the timeout for response, the query is likely to take much longer than 60 seconds.